curp/src/rpc/connect.rs

```diff
@@ -0,0 +1,267 @@
+
+    /// Send `Vote` request
+    async fn vote(
+        &self,
+        request: VoteRequest,
+        timeout: Duration,
+    ) -> Result<tonic::Response<VoteResponse>, ProposeError>;
+
+    /// Send `FetchLeaderRequest`
+    async fn fetch_leader(
+        &self,
+        request: FetchLeaderRequest,
+        timeout: Duration,
+    ) -> Result<tonic::Response<FetchLeaderResponse>, ProposeError>;
+
+    /// Convert a vec of addr string to a vec of `Connect`
+    async fn try_connect(addrs: HashMap<ServerId, String>) -> Vec<Arc<Self>>;
+
+    /// Convert a vec of addr string to a vec of `Connect`
+    #[cfg(test)]
+    async fn try_connect_test(
+        addrs: HashMap<ServerId, String>,
+        reachable: Arc<AtomicBool>,
+    ) -> Vec<Arc<Self>>;
+}
+
+/// The connection struct to hold the real rpc connections, it may failed to connect, but it also
+/// retries the next time
+#[derive(Debug)]
+pub(crate) struct Connect {
+    /// Server id
+    id: ServerId,
+    /// The rpc connection, if it fails it contains a error, otherwise the rpc client is there
+    rpc_connect: RwLock<Result<ProtocolClient<tonic::transport::Channel>, tonic::transport::Error>>,
+    /// The addr used to connect if failing met
+    addr: String,
+    /// Whether the client can reach others, useful for testings
+    #[cfg(test)]
+    reachable: Arc<AtomicBool>,
+}
+
+#[async_trait]
+impl ConnectInterface for Connect {
+    /// Get server id
+    fn id(&self) -> &ServerId {
+        &self.id
+    }
+
+    /// Get the internal rpc connection/client
+    async fn get(
+        &self,
+    ) -> Result<ProtocolClient<tonic::transport::Channel>, tonic::transport::Error> {
+        if let Ok(ref client) = *self.rpc_connect.read().await {
+            return Ok(client.clone());
+        }
+        let mut connect_write = self.rpc_connect.write().await;
+        if let Ok(ref client) = *connect_write {
+            return Ok(client.clone());
+        }
+        let client = ProtocolClient::<_>::connect(self.addr.clone())
+            .await
+            .map(|client| {
+                *connect_write = Ok(client.clone());
+                client
+            })?;
+        *connect_write = Ok(client.clone());
+        Ok(client)
+    }
+
+    /// send "propose" request
+    async fn propose(
+        &self,
+        request: ProposeRequest,
+        timeout: Duration,
+    ) -> Result<tonic::Response<ProposeResponse>, ProposeError> {
+        #[cfg(test)]
+        if !self.reachable.load(Ordering::Relaxed) {
+            return Err(ProposeError::RpcError("unreachable".to_owned()));
+        }
+
+        let mut client = self.get().await?;
+        let mut req = tonic::Request::new(request);
+        req.set_timeout(timeout);
+        req.metadata_mut().inject_current();
+        client.propose(req).await.map_err(Into::into)
+    }
+
+    /// send "wait synced" request
+    #[allow(dead_code)] // false positive report
+    async fn wait_synced(
```

Comment:
remove `#[allow(dead_code)]`, we use this in slow_round now,
and add `#[instrument(skip(self), name = "client propose")]` for tracing

Reply:
Fixed.